# Step one: Cleaning

In this notebook we will focus on taking the original data file and reducing all the errors in it. This will produce a clean file that will allow for meaningful analysis.

## Importing libraries and reading the file

In [67]:
import pandas as pd

attacks = pd.read_csv("data/attacks.csv", encoding='unicode_escape')

#We will alos run the 'cleaning.py' script as it will define functions we will use

%run -i 'src/cleaning.py'

## Initial exploration
Now we will start exploring the file. Using the ".head()" we can get a preview of the columns and a the first rows of the dataset.

In [68]:
attacks.head()

Case Number         Date    Year        Type    Country             Area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating        USA       California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  2018.06.09  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         Location     Activity             Name Sex   ...  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe    F  ...   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely     F  ...   
2                    Habush, Oahu      Surfing      John Denges    M  ...   
3              Arrawarra Headland      Surfing             male    M  ...   
4                        La Ticla  Free diving   Gustavo Ramos     M  ...   

          Species           Investigator or Source                       pdf  \
0      White shark                R. Collier, GSAF      2018.06.25-Wolfe.pdf   
1              NaN  K.McMurray, TrackingSharks.com    2018.06.18-McNeely.pdf   
2              NaN  K.McMurray, TrackingSharks.com     2018.06.09-Denges.pdf   
3        2 m shark                  B. Myatt, GSAF  2018.06.08-Arrawarra.pdf   
4  Tiger shark, 3m                       A .Kipper      2018.06.04-Ramos.pdf   

                                        href formula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href Case Number.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.25   
1  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.18   
2  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.09   
3  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.08   
4  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.04   

  Case Number.2 original order Unnamed: 22 Unnamed: 23  
0    2018.06.25         6303.0         NaN         NaN  
1    2018.06.18         6302.0         NaN         NaN  
2    2018.06.09         6301.0         NaN         NaN  
3    2018.06.08         6300.0         NaN         NaN  
4    2018.06.04         6299.0         NaN         NaN  

[5 rows x 24 columns]

In [69]:
attacks.columns

Index(['Case Number', 'Date', 'Year', 'Type', 'Country', 'Area', 'Location',
       'Activity', 'Name', 'Sex ', 'Age', 'Injury', 'Fatal (Y/N)', 'Time',
       'Species ', 'Investigator or Source', 'pdf', 'href formula', 'href',
       'Case Number.1', 'Case Number.2', 'original order', 'Unnamed: 22',
       'Unnamed: 23'],
      dtype='object')

In [70]:
attacks.shape

(25723, 24)

Using 'head()' , '.columns' and '.shape' we can see that we are dealing with a file of 24 columns (names listed above) and 25723 entries. Lets explore further:

In [71]:
attacks.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25723 entries, 0 to 25722
Data columns (total 24 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Case Number             8702 non-null   object 
 1   Date                    6302 non-null   object 
 2   Year                    6300 non-null   float64
 3   Type                    6298 non-null   object 
 4   Country                 6252 non-null   object 
 5   Area                    5847 non-null   object 
 6   Location                5762 non-null   object 
 7   Activity                5758 non-null   object 
 8   Name                    6092 non-null   object 
 9   Sex                     5737 non-null   object 
 10  Age                     3471 non-null   object 
 11  Injury                  6274 non-null   object 
 12  Fatal (Y/N)             5763 non-null   object 
 13  Time                    2948 non-null   object 
 14  Species                 3464 non-null 

From this output we gather a few things:

#### Types and names of the columns: 
* There is only two columns with floats
    * This is odd considering that the column "Age" should also be a number
    * The column "Original order" probably adds little to no value
* Columns 0, 19 and 20 all reference a case number
* Columns 17 and 18 could also be duplicated, since they both reference "href"

####  Non-null values:
* Column 0 "Case number" has significantly more non-null values than the rest of the columns
* Columns 22 and 23 are unnamed and have 2 or less non-null values, these columns add no value to the analysis
* Several columns have a non-null count nearing 6300. But most columns have less than this ammount. This means that we are dealing with a great amount of incomplete rows.


### Deleting Unnamed columns:
We will start cleaning the file by deleting the last 2 columns. For this we will use a function declared in the 'cleaning.py' ran in the beginning of this notebook:

In [72]:
delete_columns(attacks,['Unnamed: 22','Unnamed: 23'])

Deleted columns:  ['Unnamed: 22', 'Unnamed: 23']


In [73]:
attacks.sample(5)

Case Number                  Date    Year        Type  Country  \
8800            NaN                   NaN     NaN         NaN      NaN   
6640              0                   NaN     NaN         NaN      NaN   
5453   1907.07.14.R  Reported 14-Jul-1907  1907.0  Unprovoked  CROATIA   
1466   2006.06.30.a             July 2006  2006.0  Unprovoked      USA   
15302           NaN                   NaN     NaN         NaN      NaN   

                         Area                     Location      Activity  \
8800                      NaN                          NaN           NaN   
6640                      NaN                          NaN           NaN   
5453    Split-Dalmatia Count,         Su?uraj, Hvar Island      Swimming   
1466                  Florida  Florida Keys, Monroe County  Spearfishing   
15302                     NaN                          NaN           NaN   

            Name Sex   ... Fatal (Y/N) Time                     Species   \
8800         NaN  NaN  ...         NaN  NaN                          NaN   
6640         NaN  NaN  ...         NaN  NaN                          NaN   
5453      female    F  ...           Y  NaN                          NaN   
1466   Gary Hall    M  ...           N  NaN  1.8 m blacktip "reef" shark   
15302        NaN  NaN  ...         NaN  NaN                          NaN   

                        Investigator or Source                        pdf  \
8800                                       NaN                        NaN   
6640                                       NaN                        NaN   
5453   C. Moore, S. Navajas & R. Rocconi, GSAF  1907.07.14-R-Adriatic.pdf   
1466      Swimming World Magazine, 2 July 2006  2006.06.30.a-GaryHall.pdf   
15302                                      NaN                        NaN   

                                            href formula  \
8800                                                 NaN   
6640                                                 NaN   
5453   http://sharkattackfile.net/spreadsheets/pdf_di...   
1466   http://sharkattackfile.net/spreadsheets/pdf_di...   
15302                                                NaN   

                                                    href Case Number.1  \
8800                                                 NaN           NaN   
6640                                                 NaN           NaN   
5453   http://sharkattackfile.net/spreadsheets/pdf_di...  1907.07.14.R   
1466   http://sharkattackfile.net/spreadsheets/pdf_di...  2006.06.30.a   
15302                                                NaN           NaN   

      Case Number.2 original order  
8800            NaN            NaN  
6640            NaN            NaN  
5453   1907.07.14.R          850.0  
1466   2006.06.30.a         4837.0  
15302           NaN            NaN  

[5 rows x 22 columns]

### Dealing with the "Case Number" columns
Our data set contains 3 columns referring to a case number. For this we will use a function declared in the 'cleaning.py' ran in the beginning of this notebook: 

In [74]:
compare_columns(attacks,["Case Number","Case Number.1","Case Number.2"])

Column information:


Columns            Case Number  Case Number.1  Case Number.2
Lengths                  25723          25723          25723
Non - null values         8702           6302           6302

Data type comparison:


Case Number  Case Number.1  Case Number.2
<class 'float'>        17021          19421          19421
<class 'str'>           8702           6302           6302

Value comparisons:


Case Number  Case Number.1  Case Number.2
Case Number          25723          23298          23318
Case Number.1        23298          25723          25703
Case Number.2        23318          25703          25723

The above matrix shows the amount of rows that each column shares with other columns. E.g. how many cells store the same information in both columns
Sample of columns with differences:


Case Number Case Number.1 Case Number.2
17913         NaN           NaN           NaN
21697         NaN           NaN           NaN
10331         NaN           NaN           NaN
24593         NaN           NaN           NaN
2320   1996.01.23    1996.01.23    1996.01.23

We can see that the columns "Case Number.1" and "Case Number.2" hold almost the same information (They return the same values in all the tests, and in the comparison matrix they only have 20 values that are different). Lets check these values:

In [75]:
different_value_columns(attacks, ["Case Number.1", "Case Number.2"])

Columns differ in the following 20 values


Case Number.1 Case Number.2
34      2018.04.02    2018.04.03
117   2017/07.20.a  2017.07.20.a
144     2017.06.06    2017.05.06
217     2016.09.16    2016.09.15
314   2015.01.24.b  2016.01.24.b
334     2015.11.07    2015.12.23
339     2015.10.28  2015.10.28.a
560     2013.05.04    2014.05.04
3522    1967/07.05    1967.07.05
3795  1962.08.30.b  1962,08.30.b
3880  1961.09,06.R  1961.09.02.R
4403    1952.08.04    1952.08.05
4425  1951.12.15.R  1851.12.15.R
5275  1919.00.00.R  1900.00.00.R
5358  1913.17.10.R  1913.07.10.R
5386  1911.07.31.T  1911.07.31.R
5777  1884.04.28.R  1884.04.38.R
5883  1874.06.18.R  1876.06.18.R
6009    1853.04.29    1853,94.29
6122  1808.05.01.R  1808.06.01.R

We can see that the columns hold mostly the same information. Most of the differences are symbols (commas for periods or slashes) and missing letters. We will keep column Case Number.2 and drop Case Number.1

In [94]:
attacks.drop(["Case Number.1"], axis=1, inplace=True)
attacks.head()

Case Number         Date    Year        Type    Country             Area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating        USA       California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  2018.06.09  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         Location     Activity             Name Sex   ...  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe    F  ...   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely     F  ...   
2                    Habush, Oahu      Surfing      John Denges    M  ...   
3              Arrawarra Headland      Surfing             male    M  ...   
4                        La Ticla  Free diving   Gustavo Ramos     M  ...   

                                              Injury Fatal (Y/N)  \
0  No injury to occupant, outrigger canoe and pad...           N   
1                         Minor injury to left thigh           N   
2       Injury to left lower leg from surfboard skeg           N   
3                          Minor injury to lower leg           N   
4  Lacerations to leg & hand shark PROVOKED INCIDENT           N   

            Time         Species           Investigator or Source  \
0          18h00      White shark                R. Collier, GSAF   
1  14h00  -15h00              NaN  K.McMurray, TrackingSharks.com   
2          07h45              NaN  K.McMurray, TrackingSharks.com   
3            NaN        2 m shark                  B. Myatt, GSAF   
4            NaN  Tiger shark, 3m                       A .Kipper   

                        pdf  \
0      2018.06.25-Wolfe.pdf   
1    2018.06.18-McNeely.pdf   
2     2018.06.09-Denges.pdf   
3  2018.06.08-Arrawarra.pdf   
4      2018.06.04-Ramos.pdf   

                                        href formula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href Case Number.2  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.25   
1  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.18   
2  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.09   
3  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.08   
4  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.04   

  original order  
0         6303.0  
1         6302.0  
2         6301.0  
3         6300.0  
4         6299.0  

[5 rows x 21 columns]

Now lets compare "Case Number" with "Case Number.2"

In [95]:
different_value_columns(attacks, ["Case Number.2", "Case Number"])

Columns differ in the following 2405 values


Case Number.2   Case Number
390      2015.07.10    2015.07-10
4949   1934.02.08.R  1934.01.08.R
5488   1905.09.06.R              
5944     1864.05.00       1864.05
6302                            0
...             ...           ...
8698                            0
8699                            0
8700                            0
8701                            0
25722                          xx

[2405 rows x 2 columns]

We can see that "Case Number" has a great deal of 0 values, lets print the value_counts for "Case Number" where it is different from "Case Number.2"

In [98]:
print_test = different_value_columns(attacks, ["Case Number.2", "Case Number"])
print_test["Case Number"].value_counts()

Columns differ in the following 2405 values


0               2400
2015.07-10         1
1934.01.08.R       1
                   1
1864.05            1
xx                 1
Name: Case Number, dtype: int64

As we suspected, "Case Number" has 2400 values where it holds "0", one "xx" value and only 4 cases where it is similar to "Case Number.2". Since "Case Number.2" holds more complete information in these values and "Case Number" adds no value with these 0 values, we will drop it.

In [99]:
attacks.drop(["Case Number"], axis=1, inplace=True)
attacks.head()

Date    Year        Type    Country             Area  \
0  25-Jun-2018  2018.0     Boating        USA       California   
1  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         Location     Activity             Name Sex   Age  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe    F   57   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely     F   11   
2                    Habush, Oahu      Surfing      John Denges    M   48   
3              Arrawarra Headland      Surfing             male    M  NaN   
4                        La Ticla  Free diving   Gustavo Ramos     M  NaN   

                                              Injury Fatal (Y/N)  \
0  No injury to occupant, outrigger canoe and pad...           N   
1                         Minor injury to left thigh           N   
2       Injury to left lower leg from surfboard skeg           N   
3                          Minor injury to lower leg           N   
4  Lacerations to leg & hand shark PROVOKED INCIDENT           N   

            Time         Species           Investigator or Source  \
0          18h00      White shark                R. Collier, GSAF   
1  14h00  -15h00              NaN  K.McMurray, TrackingSharks.com   
2          07h45              NaN  K.McMurray, TrackingSharks.com   
3            NaN        2 m shark                  B. Myatt, GSAF   
4            NaN  Tiger shark, 3m                       A .Kipper   

                        pdf  \
0      2018.06.25-Wolfe.pdf   
1    2018.06.18-McNeely.pdf   
2     2018.06.09-Denges.pdf   
3  2018.06.08-Arrawarra.pdf   
4      2018.06.04-Ramos.pdf   

                                        href formula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href Case Number.2  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.25   
1  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.18   
2  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.09   
3  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.08   
4  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.04   

   original order  
0          6303.0  
1          6302.0  
2          6301.0  
3          6300.0  
4          6299.0